pkg/controllers/management/clusterprovisioner/provisioner.go

```diff
@@ -179,6 +192,23 @@ func (p *Provisioner) Remove(cluster *v3.Cluster) (runtime.Object, error) {
 }
 
 func (p *Provisioner) Updated(cluster *v3.Cluster) (runtime.Object, error) {
+	var err error
+
+	if shouldCleanHelmReleases(features.HelmCleanupOnUpgrade.Enabled(), cluster) {
+		cluster, err = p.startCleanUpHelmReleases(cluster)
+		if err != nil {
+			// errors here should not prevent the cluster from upgrading
+			// log and continue the process
+			logrus.Errorf("failed to clean up Helm releases in upgraded cluster: %#v", err)
+		}
+
+		cluster, err = p.Clusters.Get(cluster.Name, metav1.GetOptions{})
+		if err != nil {
+			// if we failed to get a cluster, something bad must have happened
+			return cluster, errors.Wrapf(err, "[updated] failed to get updated cluster %#v after startCleanUpHelmReleases", cluster.Name)
+		}
```

Comment:
you'll want to do this in the function that <condition>.Do wraps. The reason why is because the Do() function does not have access to a client and therefore is only updating the cluster object for you to pass to the client's Update function. Right now, everything done above is just being overwritten by what's already in etcd.

Reply:
Ah, that's a bit confusing. I'll change it.

Reply:
See [here](https://github.com/rancher/rancher/pull/40104/files#r1070106466).

Reply:
Updated!